metrics_sdk/lib/opentelemetry/sdk/metrics/aggregation/sum.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Metrics
+      module Aggregation
+        # Contains the implementation of the Sum aggregation
+        # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#sum-aggregation
+        class Sum
+          def initialize
+            @aggregation_temporality = :delta
+
+            @data_points = {}
+          end
+
+          def collect(start_time, end_time)
+            @data_points.each_value do |ndp|
+              ndp.start_time_unix_nano = start_time
+              ndp.time_unix_nano = end_time
+            end
+            ndps = @data_points.values
+            if @aggregation_temporality == :delta
+              @data_points.clear
+            else
+              ndps = ndps.map(&:dup)
+            end
+            ndps
```

Comment:
The logic here is tricky. I think we need to distinguish `@aggregation_temporality == :delta` at the top-level and handle the two cases separately. Otherwise there'll be a bunch of messy conditionals. Here's my attempt:
```suggestion
            if @aggregation_temporality == :delta
              ndps = @data_points.map do |_, ndp|
                ndp.start_time_unix_nano = start_time
                ndp.time_unix_nano = end_time
                ndp
              end
              @data_points.clear
              ndps
            else
              @data_points.map do |_, ndp|
                ndp.start_time_unix_nano ||= start_time
                ndp.time_unix_nano = end_time
                ndp.dup
              end
            end
```
The main difference is that we don't need to `dup` the `ndp` in the `delta` case because we `clear` the hash, and in the non-`delta` case we need to only set the `start_time` once for each `ndp` *and then* `dup`. Unless I misunderstand aggregation temporality, non-`delta` data points should have a `start_time` of when they were first observed.

Reply:
Alternative:
```suggestion
            if @aggregation_temporality == :delta
              # Set timestamps and 'move' data point values to result.
              ndps = @data_points.values.each do |ndp|
                ndp.start_time_unix_nano = start_time
                ndp.time_unix_nano = end_time
              end
              @data_points.clear
              ndps
            else
              # Update timestamps and take a snapshot.
              @data_points.values.map! do |ndp|
                ndp.start_time_unix_nano ||= start_time # Start time of a data point is from the first observation.
                ndp.time_unix_nano = end_time
                ndp.dup
              end
            end
```